feat(curriculum): added the Tower of Hanoi Python project (#51347)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/20-upcoming-python/learn-recursion-by-solving-the-tower-of-hanoi-puzzle/64de7be06eb689161dd63cf0.md

---
id: 64de7be06eb689161dd63cf0
title: Step 36
challengeType: 20
dashedName: step-36
---

# --description--

The conditionals you previously wrote apply only for odd number of disks. Add a nested `if` to execute when `n` is odd and indent of one level your `print` and `make_allowed_move` calls.
# --hints--

> You should nest an `if` statement to execute when `n` is odd. Don't forget to indent of one level your `print` and `make_allowed_move` calls.

Reviewer: ```suggestion
You should nest an `if` statement to execute when `n` is odd. Don't forget to indent your `print` and `make_allowed_move` calls one more level.
```

---
id: 64de7be06eb689161dd63cf0
title: Step 38
challengeType: 20
dashedName: step-38
---

# --description--

The conditionals you previously wrote apply only for odd number of disks. Add a nested `if` to execute when `n` is odd, and add one indent level to your `print` and `make_allowed_move` calls.
# --hints--

You should nest an `if` statement to execute when `n` is odd. Don't forget to indent your `print` and `make_allowed_move` calls one more level.

```js

> ({ test: () => assert.match(code, /if\s+n\s*%\s*2\s*!=\s*0\s*:\s+print.*\s+make_allowed_move\(.*\)/) })

Reviewer 2: ```suggestion
({ test: () => assert.match(code, /if\s+n\s*%\s*2\s*(!=\s*0|==\s*1)\s*:\s+print.*\s+make_allowed_move\(.*\)/) })
```
Alternative version.

```

# --seed--

## --seed-contents--

```py
NUMBER_OF_DISKS = 4
number_of_moves = 2 ** NUMBER_OF_DISKS - 1
rods = {
    'A': list(range(NUMBER_OF_DISKS, 0, -1)),
    'B': [],
    'C': []
}

def make_allowed_move(rod1, rod2):    
    forward = False
    if not rods[rod2]:
        forward = True
    elif rods[rod1] and rods[rod2] and rods[rod1][-1] < rods[rod2][-1]:
        forward = True              
    if forward:
        print(f'Moving disk {rods[rod1][-1]} from {rod1} to {rod2}')
        rods[rod2].append(rods[rod1].pop())
    else:
        print(f'Moving disk {rods[rod2][-1]} from {rod2} to {rod1}')
        rods[rod1].append(rods[rod2].pop())
    
    # display our progress
    print(rods, '\n')

def move(n, source, target, auxiliary):
    # display starting configuration

>     print(rods, '\n')   

Reviewer 2: ```suggestion
    print(rods, '\n')
```
And further steps

---
id: 64de7be06eb689161dd63cf0
title: Step 38
challengeType: 20
dashedName: step-38
---

# --description--

> The conditionals you previously wrote apply only for odd number of disks.

Reviewer 3: ```suggestion
The conditionals you wrote previously are only valid for odd numbers of disks.
```

Add a nested `if` to execute when `n` is odd, and add one indent level to your `print()` and `make_allowed_move()` calls.

# --hints--

You should nest an `if` statement to execute when `n` is odd. Don't forget to indent your `print()` and `make_allowed_move()` calls.

```js
({ test: () => assert.match(code, /if\s+n\s*%\s*2\s*(!=\s*0|==\s*1)\s*:\s+print.*\s+make_allowed_move\(.*\)/) })
```

# --seed--

## --seed-contents--

```py
NUMBER_OF_DISKS = 4

> number_of_moves = 2 ** NUMBER_OF_DISKS - 1

Reviewer 3: ```suggestion
number_of_moves = 2**NUMBER_OF_DISKS - 1
```

rods = {
    'A': list(range(NUMBER_OF_DISKS, 0, -1)),
    'B': [],
    'C': []
}

def make_allowed_move(rod1, rod2):    
    forward = False
    if not rods[rod2]:
        forward = True
    elif rods[rod1] and rods[rod1][-1] < rods[rod2][-1]:

>         forward = True              
>     if forward:

Reviewer 3: ```suggestion
        forward = True
                      
    if forward:
```

        print(f'Moving disk {rods[rod1][-1]} from {rod1} to {rod2}')
        rods[rod2].append(rods[rod1].pop())
    else:
        print(f'Moving disk {rods[rod2][-1]} from {rod2} to {rod1}')
        rods[rod1].append(rods[rod2].pop())
    
    # display our progress
    print(rods, '\n')

def move(n, source, auxiliary, target):

>     # display starting configuration

Reviewer 3: ```suggestion
    # Display starting configuration
```

    print(rods, '\n')
    for i in range(number_of_moves):
        remainder = (i + 1) % 3
--fcc-editable-region--     
        if remainder == 1:
            
            print(f'Move {i + 1} allowed between {source} and {target}')
            make_allowed_move(source, target)
--fcc-editable-region--            
        elif remainder == 2:
            print(f'Move {i + 1} allowed between {source} and {auxiliary}')
            make_allowed_move(source, auxiliary)
        elif remainder == 0:
            print(f'Move {i + 1} allowed between {auxiliary} and {target}')
            make_allowed_move(auxiliary, target)
           

> # initiate call from source A to target C with auxiliary B

Reviewer 3: ```suggestion
# Initiate call from source A to target C with auxiliary B
```
